lilypad/closure.py

```diff
@@ -0,0 +1,543 @@
+                continue
+            spec = importlib.util.find_spec(root_module)
+            if spec is None or spec.origin is None:  # pragma: no cover
+                continue
+            dist = importlib.metadata.distribution(root_module)
+
+            if dist:
+                extras = []
+                for extra in dist.metadata.get_all("Provides-Extra", []):
+                    extra_reqs = dist.requires or []
+                    extra_deps = [
+                        Requirement(r).name
+                        for r in extra_reqs
+                        if f"extra == '{extra}'" in r
+                    ]
+                    if extra_deps and all(
+                        dep in installed_packages for dep in extra_deps
+                    ):
+                        extras.append(extra)
+
+                dependencies[dist.name] = {
+                    "version": dist.version,
+                    "extras": extras if extras else None,
+                }
+
+        return dependencies
+
+    def collect(
+        self, fn: Callable[..., Any]
+    ) -> tuple[list[str], list[str], list[str], dict[str, DependencyInfo]]:
+        """Returns the imports and source code for a function and its dependencies."""
+        self._collect_imports_and_source_code(fn, True)
+
+        local_names = set()
+        for code in self.source_code + self.assignments:
+            tree = ast.parse(code)
+            for node in ast.walk(tree):
+                if isinstance(node, ast.FunctionDef | ast.ClassDef):
+                    local_names.add(node.name)
+
+        rewriter = _QualifiedNameRewriter(local_names, self.user_defined_imports)
+
+        assignments = []
+        for code in self.assignments:
+            tree = ast.parse(code)
+            new_tree = rewriter.visit(tree)
+            assignments.append(ast.unparse(new_tree))
+
+        source_code = []
+        for code in self.source_code:
+            tree = ast.parse(code)
+            new_tree = rewriter.visit(tree)
+            source_code.append(ast.unparse(new_tree))
+
+        required_dependencies = self._collect_required_dependencies(
+            self.imports | self.fn_internal_imports
+        )
+
+        return (
+            list(self.imports),
+            list(dict.fromkeys(assignments)),
+            source_code,
+            required_dependencies,
+        )
+
+
+def _run_ruff(code: str) -> str:
+    with tempfile.NamedTemporaryFile(mode="w", suffix=".py", delete=False) as tmp_file:
+        tmp_file.write(code)
+        tmp_path = Path(tmp_file.name)
+
+    try:
+        subprocess.run(
+            ["ruff", "check", "--select=I", "--fix", str(tmp_path)],
+            check=True,
+            capture_output=True,
+            text=True,
+        )
+        subprocess.run(
+            ["ruff", "format", "--line-length=88", str(tmp_path)],
+            check=True,
+            capture_output=True,
+            text=True,
+        )
```

Comment:
Although I can’t think of a concrete scenario, there is a possibility that the generated code could change if ruff detects configuration files like ruff.toml or pyproject.toml.
Should we specify certain options to ensure the results remain consistent?
If this concern is not relevant, feel free to ignore this comment.